assets/src/edit-story/utils/useUploadVideoFrame.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * WordPress dependencies
+ */
+import { useCallback } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import { useAPI } from '../app/api';
+import { useStory } from '../app/story';
+import { useConfig } from '../app/config';
+import getFirstFrameOfVideo from './getFirstFrameOfVideo';
+
+function useUploadVideoFrame( { videoId, src, id } ) {
+	const { actions: { uploadMedia, saveMedia } } = useAPI();
+	const { storyId } = useConfig();
+	const { actions: { updateElementById } } = useStory();
+	const setProperties = useCallback(
+		( properties ) => updateElementById( { elementId: id, properties } ),
+		[ id, updateElementById ] );
+
+	const processData = async () => {
+		try {
+			const obj = await getFirstFrameOfVideo( src );
+			const { id: featuredMedia, source_url: featuredMediaSrc } = await uploadMedia( obj );
+			await saveMedia( featuredMedia, {
+				meta: {
+					_web_stories_is_poster: true,
+				},
+				post: storyId,
+			} );
+			await saveMedia( videoId, {
+				featured_media: featuredMedia,
+				post: storyId,
+			} );
+			const newState = { featuredMedia, featuredMediaSrc };
+			setProperties( newState );
+		} catch ( err ) {
+			// TODO Display error message to user as video poster upload has as failed.
+		}
+	};
+
+	/**
+	 * Uploads the video's first frame as an attachment.
+	 *
+	 */
+	const uploadVideoFrame = useCallback( processData, [ getFirstFrameOfVideo, src, uploadMedia, saveMedia, videoId, setProperties ] );
+
+	return {
+		uploadVideoFrame,
+	};
+}
```

Comment:
This function can be tested by mocking out all of `useAPI`, `useConfig`, `useStory` and `getFirstFrameOfVideo`. But it's a lot of work - I can't really tell if it's worth it. Of course there are a lot of error cases, that could be tested for, but as they aren't really handled yet, testing them doesn't make a lot of sense.

Reply:
Agreed, feel like more work than the value it would add.